edweiss-firebase/functions/src/functions/lectures/likeQuestion.ts

```diff
@@ -1,31 +1,41 @@
+import { Timestamp } from 'firebase-admin/firestore';
 import LectureDisplay from 'model/lectures/lectureDoc';
 import { onAuthentifiedCall } from 'utils/firebase';
-import { clean, CollectionOf, getDocumentAndRef } from 'utils/firestore';
-import { fail, ok } from 'utils/status';
+import { CollectionOf, getDocumentAndRef } from 'utils/firestore';
+import { fail, OK } from 'utils/status';
 
-export const updateQuestion = onAuthentifiedCall(LectureDisplay.Functions.likeQuestion, async (userId, args) => {
+export const likeQuestion = onAuthentifiedCall(LectureDisplay.Functions.likeQuestion, async (userId, args) => {
     if (!args.id || args.id.length == 0)
         return fail("invalid_id");
 
-    const [question, ref] = await getDocumentAndRef(CollectionOf<LectureDisplay.Question>(`courses/${args.courseId}/lectures/${args.lectureId}/questions`), args.id);
-
-    if (question == undefined)
-        return fail("error_firebase");
-
-    const updatedQuestion: Partial<LectureDisplay.Question> = {
-        anonym: question.anonym,
-        likes: args.likes,
-        postedTime: question.postedTime,
-        text: question.text,
-        userID: question.userID,
-        username: question.username
-    };
-    try {
-        console.log(updateQuestion);
-        await ref.update(clean(updatedQuestion))
-    } catch (error) {
-        return fail('error_firebase');
-    }
+    const [question, questionRef] = await getDocumentAndRef(CollectionOf<LectureDisplay.Question>(`courses/${args.courseId}/lectures/${args.lectureId}/questions`), args.id);
+    const [like, likeRef] = await getDocumentAndRef(CollectionOf<LectureDisplay.Like>(`courses/${args.courseId}/lectures/${args.lectureId}/questions/${args.id}/likes`), userId);
+
+    if (args.liked) {
+        if (like != undefined)
+            return OK;
+
+        await likeRef.set({
+            createdAt: Timestamp.now()
+        });
+
+        await questionRef.set({
+            ...question,
+            likes: question.likes + 1
+        });
 
-    return ok({ id: ref.id });
-});
\ No newline at end of file
+        return OK;
+    } else {
+        if (like == undefined)
+            return OK;
+
+        await likeRef.delete();
+
+        await questionRef.set({
+            ...question,
+            likes: Math.max(question.likes - 1, 0)
+        });
```

Comment:
I appreciate your refactoring it makes total sens. You did a great job at using the await for the delete so that we can update correctly the reference.
Furthermore, you did a great refactoring in this file, it much cleaner and understanble from an external standpoint.

Reply:
Yes we do not want values lower then 0 for the likes as well